File: leetcode/algorithms/hash-table/longest-substring-wihout-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer> charMap = new HashMap<Character,Integer>();
        
        int i = 0, len = 0,maxLen = 0;
        while(i<s.length()){
            char currChar = s.charAt(i);
            Integer index = charMap.getOrDefault(currChar,null);
            
            if(index!=null){
                i = index+1;
                charMap.clear();
                maxLen = Math.max(maxLen,len);
                len = 0;
            } else {
                charMap.put(currChar,i);
                ++i;
                ++len;
                maxLen = Math.max(maxLen,len);
            }
        }
        return maxLen;
    }
}

File: leetcode/algorithms/hash-table/longest-substring-wihout-repeating-characters.ts
function lengthOfLongestSubstring(s: string): number {
    
    const charIndexMap = new Map<string,number>();
    let maxLen = 0,currStartIndex = 0;
    for(let i = 0;i<s.length;++i){
        const ch: string = s.charAt(i);
        const currCharIndex = charIndexMap.get(ch) ?? null;
        if(currCharIndex!==null && currCharIndex>=currStartIndex){
            currStartIndex = currCharIndex + 1;
        } else {
            maxLen = Math.max(maxLen,i - currStartIndex + 1);
        }
        charIndexMap.set(ch,i);
    }
    
    return maxLen;
    
};